guides/v2.1/coding-standards/docblock-standard-general.md

```diff
@@ -326,27 +328,29 @@ class Profiler
 ### Functions and methods
 {:#functions-methods}
 
-Functions and methods must have:
+In general, typed method signature should be preferred over PHPDoc annotations when possible.
 
-* Short description
-* Long description that explains the motivation behind the implementation.
+Functions and methods should have:
+
+* A short description in case it adds meaningful information beyond the method name.
+* If the purpose of the method is not obvious, a long description that explains the motivation behind the implementation.
   For example:
 
    * If a workaround or hack is implemented, explain why it is necessary and include any other details necessary to understand the algorithm.
    * For non-obvious implementations where the implementation logic is complicated or does not correspond to the Technical Vision or other known best practices, include an explanation in the doc block's description.
      An implementation is non-obvious if another developer has questions about it.
 
-* Declaration of all arguments (if any) using `@param` tag.
-  Appropriate argument type must be specified.
-* Declaration of return type using `@return` tag.
-  If there is no such operator, the `@return` tag must have `void` as the return value.
+* The declaration of all arguments (if any) using `@param` tag, unless the argument type is indicated in the method signature.
+  All `@param` annotations must include the appropriate argument type.
+  If any argument requires a `@param` annotation, all arguments must be listed (all or none).
```

Comment:
Disagree. Let's specify `@param` only for necessary arguments, when it is more accurate than type hinting.

Reply:
In my experience that leads to less readable code.

Reply:
Some examples? I don't think `@param` annotations are for humans at all, they are to give maximum accurate information for IDE. Besides constructor, any method don't contain more than 4-5 parameters usually, so, I believe such block with partial types for arguments missing real type hinting will be readable anyway.

Reply:
I know default PhpStorm will complain about partial doc blocks. We could change this, but a lot of developers will not understand it and it will lead to inconsistencies. Not sure how IDEs or PhpDoc will handle the order of the arguments then.

Reply:
I don't think there is an absolute truth to be reached in this regard (as is often the case in regards to "understandability". Who can decide this matter?

Reply:
Can we also add a rule that `@param` annotations should be in the same order as method arguments?

Reply:
@lenaorobei done

Reply:
@Vinai,

> In my experience incomplete lists of params in the annotations lead to less readable code.

Do you have some examples? There are not a lot of situations when we will need annotation for parameter at all (only more accurate type like `array` -> `Product[]` and such). Methods beside constructor usually won't contain more than 3-4 arguments.

Reply:
I don't have any examples at hand, since I changed those to contain all parameters instead.
But here is a constructed example:

Single `@param` annotation:
```php
/**
 * @param SectionSourceInterface[] $sectionSourceMap
 */
public function __construct(
    ObjectManagerInterface $objectManager,
    SectionIdentifier $identifier,
    array $sectionSourceMap = []
) {
    $this->objectManager = $objectManager;
    $this->identifier = $identifier;
    $this->sectionSourceMap = $sectionSourceMap;
}
```

A complete list of parameters annotations:
```php
/**
 * @param ObjectManagerInterface $objectManager
 * @param SectionIdentifier $identifier
 * @param SectionSourceInterface[] $sectionSourceMap
 */
public function __construct(
    ObjectManagerInterface $objectManager,
    SectionIdentifier $identifier,
    array $sectionSourceMap = []
) {
    $this->objectManager = $objectManager;
    $this->identifier = $identifier;
    $this->sectionSourceMap = $sectionSourceMap;
}
```